### PR TITLE
[RFC, WIP] partial conversion to structured errors

### PR DESCRIPTION
**Pull request by tanriol**
An alternative to #4 in the part of errors. Just a draft for now to check the idea.

### CODE REVIEW

**Comment by tanriol** [comment]
The two parts not yet converted are `Read`/`Write` impls (these are libusb errors internally) and `usb_open` (looks underdocumented).

File: src/error.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    #[error("the device has been disconnected from the system")]
+    Disconnected,
+    #[error("the device does not have the specified interface")]
+    NoSuchInterface,
+    #[error("libftdi reported error to perform operation")]
+    RequestFailed,
+    #[error("input value invalid: {0}")]
+    InvalidInput(&'static str),
+
+    #[error("unknown or unexpected libftdi error: {message}")]
```
Comment (yaahc) [inline src/error.rs:20]: I would change this to introduce a source error to represent the FTDI error and not include the msg in the `Display` output here to allow error reporters to more flexibly format the error report

Comment (tanriol) [reply src/error.rs:20]: By source error do you mean introducing a separate struct for this variant?

In other situation I'd read this as "wrap the original type implementing `std::error::Error`", but there's no original type here. When `libftdi` reports an error, the two pieces of information we have are error codes (small negative integers, meaning depends on the function returning them) and error messages (strings, not documented anywhere except by source code itself).

So the common problem we're trying to solve with @cr1901 is that there's no good source error. We can preserve the message and/or the `(function_name, error_code)` tuple, and that's all. And, by the way, the `(function_name, error_code)` tuple does not uniquely identify a message in some functions (and the message does not uniquely identify either `function_name` or `function_code`).

Comment (yaahc) [reply src/error.rs:20]: > By source error do you mean introducing a separate struct for this variant?

yes, something like this

```rust
use std::fmt;

#[derive(Debug)]
struct ForeignError {
    msg: &'static str,
}

impl fmt::Display for ForeignError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.msg)
    }
}

impl std::error::Error for ForeignError {}

impl Error {
    pub(crate) fn unknown(context: *mut ffi::ftdi_context) -> Self {
        let msg = unsafe { CStr::from_ptr(ffi::ftdi_get_error_string(context)) }
            .to_str()
            .expect("all error strings are expected to be ASCII");
        let source = ForeignError { msg };
        Error::Unknown { source }
    }
}
```

or maybe even

```rust
use std::fmt;

#[derive(Debug)]
struct ForeignError {
    context: *mut ffi::ftdi_context,
}

impl fmt::Display for ForeignError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let msg = unsafe { CStr::from_ptr(ffi::ftdi_get_error_string(self.context)) }
            .to_str()
            .expect("all error strings are expected to be ASCII");
        f.write_str(msg)
    }
}

impl std::error::Error for ForeignError {}

impl Error {
    pub(crate) fn unknown(context: *mut ffi::ftdi_context) -> Self {
        let source = ForeignError { context };
        Error::Unknown { source }
    }
}
```

> So the common problem we're trying to solve with @cr1901 is that there's no good source error. We can preserve the message and/or the `(function_name, error_code)` tuple, and that's all. And, by the way, the `(function_name, error_code)` tuple does not uniquely identify a message in some functions (and the message does not uniquely identify either `function_name` or `error_code`).

I guess my question is, what level of detail do you need? Do you expect users to need to handle these errors and recover from them or only to report them? I've been mostly focused on the latter, which I think is a good place to start given that it has the minimal API surface that you would later have to consider breaking.

If you need to be able to match upon the foreign error types then I would probably just store the error code for foreign errors and export it via a member function that can be accessed in cases where the user needs to. And if the mappings between functions being different is an issue then I would add multiple variants to handle Unknown each which corresponds where the foreign error came from.

Comment (tanriol) [reply src/error.rs:20]: > I guess my question is, what level of detail do you need?

The main (and maybe only) use case for this variant is as a catch-all for "libftdi returned an undocumented error" or "this error code is so poorly documented that I have no idea how to classify it". The idea is to separate the errors that can be reasonably matched into different variants and leave this one for the "everything else" group that cannot be reasonably matched. If a user can explain what this error means and wants to match it, they should submit an issue/PR and move it to the proper variant, possibly a new one. So no, this variant is supposed to be reported, not handled.

Thank you for the suggestions!

**Comment by cr1901** [comment]
@tanriol Let's go with your solution and document the possible variants each function in `ftdi-rs` can return. I would still like an error specific to `LibUSB` in the same vein as `ForeignError` (perhaps the `Error::LibUsb` variant contains a `LibUsbError` `enum`?). But as soon as you've merged this, I'll pull it into my branch and we can move on to porting more functions over.

Thank you @yaahc for the feedback on _both_ our error handling solutions and all-around Rust help :D!

**Comment by yaahc** [comment]
My pleasure, I'm glad I was able to help!

**Pull request merged**